# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [27]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [28]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [29]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [30]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [31]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [32]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [33]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [34]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [35]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [36]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [37]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [38]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [39]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [40]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [41]:
data.duplicated().sum()

71

In [42]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [43]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [44]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [45]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [46]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [47]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [48]:
data_child_debt=data.pivot_table(index=['children'], values='debt', aggfunc=['sum','count','mean'])# сводная таблица
data_child_debt=data_child_debt.droplevel(1, axis=1).reset_index()
print(data_child_debt)

   children   sum  count      mean
0         0  1063  14091  0.075438
1         1   444   4808  0.092346
2         2   194   2052  0.094542
3         3    27    330  0.081818
4         4     4     41  0.097561
5         5     0      9  0.000000


<div class="alert alert-block alert-danger">
<b>Комментарий ревьюер:</b> 

❌ Стоит каждую таблицу называть своим именем, вдруг тебе понадобятся ее еще раз вывести, название каждой таблицы должно быть индивидуальным)) 

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Здесь молодец, отличную таблицу построил, но для красоты лучше избавляться от мультииндексов, вначале срезаем один слой названия с помощью `droplevel(1, axis=1)`, а затем сбрасываешь индексы с помощью `reset_index()`)) Это касается изучение всех взаимосвязей))  <br>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  Здесь все отлично стало)) Все замечания исправил)) Толькот с помощью `Print()`  не нужно выводить таблицу, просто название прописываешь таблицы и все выведется 

  </div>

**Вывод: люди, которые не имеют детей, лучше всего отдают кредит, доля должников здесь составляет с 7,5%, хуже обстоит дело с должниками, кто имеет одного или двух детей, здесь доля невозврата составляет соответственно 9,2% и 9,4%. Остальные категории уже некорректно сравнивать, так как в них очень маленькая выборка. В итоге можно сказать следующее: повышение количества детей увеличивает вероятность не возврата кредита.** 

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Не бойся использовать в выводах цифры, также нужно сделать акцент на сбалансированности выборок. Примерно выводы будут выглядеть так: 
`люди, которые не имеют детей, лучше всего отдают кредит, доля должников здесь составляет с 7,5%, хуже обстоит дело с должниками, кто имеет одного или двух детей, здесь доля невозврата составляет соответственно 9,2% и 9,4%. Остальные категории уже некорректно сравнивать, так как в них очень маленькая выборка`)) <br>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  Выводы переработал, молодец)) 

  </div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [49]:
data_family_status_debt=data.pivot_table(index=['family_status'], values='debt', aggfunc=['sum','count','mean'])
data_family_status_debt=data_family_status_debt.droplevel(1, axis=1).reset_index()
print(data_family_status_debt)

           family_status  sum  count      mean
0  Не женат / не замужем  273   2796  0.097639
1              в разводе   84   1189  0.070648
2         вдовец / вдова   63    951  0.066246
3       гражданский брак  385   4134  0.093130
4        женат / замужем  927  12261  0.075606


**Вывод: вероятность возникновения задолжности (mean debt) имеет максимальные показатели  у групп людей: Не женат / не замужем и гражданский брак (9,7% и 9,3% соответственно). Минимальные показатели наблюдаются у группы вдовец / вдова (6,6%). Если же говорить о максимальном количестве людей с задолжностью, то тут в лидерах категория женат / замужем, но в тоже время доля выданных кредитов в этой категории максимальна. Минимальное количество людей с задолжностью в группе вдовец / вдова, но тут и количество кредитов минимальное.
Общий вывод такой сеймейные люди имеют средние показатели по возникновениям задолжности, наиболее "опасной" категорией людей являются: Не женат / не замужем и гражданский брак.** 

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ В выводах давай будет использовать цифры, также  удели здесь внимание тоже сбалансированности выборок))

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  Здесь тоже все отлично по выводам))

  </div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [50]:
data_total_income_debt=data.pivot_table(index=['total_income_category'], values='debt', aggfunc=['sum','count','mean'])
data_total_income_debt=data_total_income_debt.droplevel(1, axis=1).reset_index()
print(data_total_income_debt)

  total_income_category   sum  count      mean
0                     A     2     25  0.080000
1                     B   354   5014  0.070602
2                     C  1353  15921  0.084982
3                     D    21    349  0.060172
4                     E     2     22  0.090909


**Вывод: 
0–30000 — 'E';
30001–50000 — 'D';
50001–200000 — 'C';
200001–1000000 — 'B';
1000001 и выше — 'A'
Категории A, D и E имеют малую выборку, поэтому их сравнивать не корректно. Среди двух оставшихъся лидером по вероятности появления задолжности является категория C - 8,4%. Также в этой категории максимальные показатели по количеству выданных кредитов и возникших задолжностей. ** 

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюер:</b> 

⚠️ Здесь сравниваем только две группы B и С, не забываем про сбалансированость выборки))<br>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️  Здесь сравнил две самые большие группы))

  </div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [51]:
data_purpose_category_debt=data.pivot_table(index=['purpose_category'], values='debt', aggfunc=['sum','count','mean'])
data_purpose_category_debt=data_purpose_category_debt.droplevel(1, axis=1).reset_index()
print(data_purpose_category_debt)


           purpose_category  sum  count      mean
0    операции с автомобилем  400   4279  0.093480
1  операции с недвижимостью  780  10751  0.072551
2     получение образования  369   3988  0.092528
3        проведение свадьбы  183   2313  0.079118


**Вывод:по вероятности задолжности наибольшим значением обладает категория операций с автомобилем (9,3%), наименьшей - операции с недвижимостью (7,2%), также в этой категории максимальные показатели по колличеству выданных кредитов и колличеству должников (780 и 10751 соответственно). Скорее всего в категории операции с недвижимостью превалируют кредиты по ипотеке, с залогом в виде приобретаемого имущества. Вероятность остаться без жилья стимулирует многих, но не всех.** 

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b> 

⚠️ Почему по твоему самые надежные заемщики те, кто берет ипотеку? Здесь можно порасуждать в выводах)) 

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Логика твоя верная, но здесь можно просто порасуждать, самые большие суммы при ипотеке, поэтому банки предъявляют самые жесткие требования к ипотечникам, поэтому их проверяют вдоль и поперек при выдаче кредитов, плюс собственность находится в залоге у банка, не будешь платить она вернется к банку)) 
   
       
  </div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ: общая проблема заключается в вариативности при заполненнии датасета. И в том числе в возможности не предоставлять данные. Решение: жестко обязать заполнять данные в соответствии с шаблоном или выбирать из списка.* 

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b>

⚠️ Однако, если пристально взглянуть на на наши данные, то увидим что пропуски в столбцах "трудовой стаж" и "ежемесячный доход" абсолютно идентичны. Количество ячеек с отсутствующими данными в обоих указанных столбцах совпадает и равно 2174. Почему такое могло возникнуть?

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>

✔️  Маловероятно, что специалисты по выдаче кредитов, заполнив все без исключения ячейки других столбцов по каждому кредитополучателю, оставляли незаполненными именно два показателя, и сделали это ровно 2174 раза в каждом столбце. Даже не проводя проверку, во всех ли случаях незаполненные ячейки находятся в одной и то же строке с высокой долей вероятности можно утверждать, что произошла техническая ошибка при формировании таблицы с исходными данными (либо некий сбой во время выгрузки данных на анализ, либо при объединении данных в единый файл, который также должен был передаваться на анализ, либо эти данные не собирались).
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства.* 

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>

✔️ Ответ правильный)) 
  </div>

### Шаг 4: общий вывод.

1.люди, которые не имеют детей, лучше всего отдают кредит, доля должников здесь составляет с 7,5%, хуже обстоит дело с должниками, кто имеет одного или двух детей, здесь доля невозврата составляет соответственно 9,2% и 9,4%. Остальные категории уже некорректно сравнивать, так как в них очень маленькая выборка. В итоге можно сказать следующее: повышение количества детей увеличивает вероятность не возврата кредита;
2. вероятность возникновения задолжности (mean debt) имеет максимальные показатели  у групп людей: Не женат / не замужем и гражданский брак (9,7% и 9,3% соответственно). Минимальные показатели наблюдаются у группы вдовец / вдова (6,6%). Если же говорить о максимальном количестве людей с задолжностью, то тут в лидерах категория женат / замужем, но в тоже время доля выданных кредитов в этой категории максимальна. Минимальное количество людей с задолжностью в группе вдовец / вдова, но тут и количество кредитов минимальное.
Общий вывод такой сеймейные люди имеют средние показатели по возникновениям задолжности, наиболее "опасной" категорией людей являются: Не женат / не замужем и гражданский брак;
3. Категории A, D и E имеют малую выборку, поэтому их сравнивать не корректно. Среди двух оставшихъся лидером по вероятности появления задолжности является категория C - 8,4%. Также в этой категории максимальные показатели по количеству выданных кредитов и возникших задолжностей;
4. по вероятности задолжности наибольшим значением обладает категория операций с автомобилем (9,3%), наименьшей - операции с недвижимостью (7,2%), также в этой категории максимальные показатели по колличеству выданных кредитов и колличеству должников (780 и 10751 соответственно). Скорее всего в категории операции с недвижимостью превалируют кредиты по ипотеке, с залогом в виде приобретаемого имущества. Вероятность остаться без жилья стимулирует многих, но не всех.
В данных много пропусков. общая проблема заключается в вариативности при заполненнии датасета. И в том числе в возможности не предоставлять данные. Решение: жестко обязать заполнять данные в соответствии с шаблоном или выбирать из списка.



 <div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>

✔️ Выводы хорошо переработала, молодец)) Рекомендации даны банку, исследование проведено))<br>
Но одним текстом тяжело читается выводы, надо разбивать все по абзацам, вот так вот:
- Вывод про детей
- Вывод про семейное положение
- Так далее

    
    
Я дам тебе ссылки на полезные гайды, которые могут помочь:
    
<a href="https://markdown-guide.readthedocs.io/en/latest/basics.html">guide1.</a>
<a href="https://paulradzkov.com/2014/markdown_cheatsheet/">guide2.</a>
    
А также советую тебе взглянуть на <a href="https://www.kaggle.com/notebooks/">работы</a> профессиональных аналитиков, чтобы оценить, как другие люди оформляют и проводят исследования.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера:</b> 

⚠️ Выводы у тебя обновятся, но в целом мне понравилось, что прям детальные сделал, добавь туда только цифр и акцент на сбалансированости сделай))

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера:</b>
    
Сергей, отлично поработал, видно что умеешь работать с кодом, в общем молодец, но нужно будет оптимизировать код при изучении взаимосвязей и поработать над выводами)) 
<br> Жду исправленного варианта)
  </div>

Добрый день. Вроде учел замечания.

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_V2:</b>
    
Сергей, отлично поработал, можно двигаться дальше)) Успехов в дальнейшем обучении)) 
  </div>